network/benchmark/src/lib.rs

```diff
@@ -0,0 +1,605 @@
+                    Handle::current().spawn(rpc_sender(
+                        node_config.clone(),
+                        network_client.clone(),
+                        time_service.clone(),
+                        network_id,
+                        wat.remote_peer_id,
+                        shared.clone(),
+                    ));
+                }
+            },
+            Event::LostPeer(_) => {}, // don't care
+        }
+    }
+}
+
+/// run_benchmark_service() does not return, it should be called by .spawn()
+pub async fn run_benchmark_service(
+    node_config: NodeConfig,
+    network_client: NetworkClient<BenchmarkMessage>,
+    network_requests: NetworkServiceEvents<BenchmarkMessage>,
+    time_service: TimeService,
+    shared: Arc<RwLock<BenchmarkSharedState>>,
+) {
+    let config = node_config.benchmark.unwrap();
+    let benchmark_service_threads = config.benchmark_service_threads;
+    let num_threads = match benchmark_service_threads {
+        Some(x) => x,
+        None => match std::thread::available_parallelism() {
+            Ok(val) => val.get(),
+            Err(_) => 1,
+        },
+    };
+    let (work_sender, work_receiver) = async_channel::bounded(num_threads * 2);
+    let runtime_handle = Handle::current();
+    let source_thread = runtime_handle.spawn(source_loop(network_requests, work_sender));
+    let mut handlers = vec![];
+    for _ in 0..num_threads {
+        handlers.push(runtime_handle.spawn(handler_thread(
+            node_config.clone(),
+            network_client.clone(),
+            work_receiver.clone(),
+            time_service.clone(),
+            shared.clone(),
+        )));
+    }
+    if let Err(err) = source_thread.await {
+        warn!("benchmark source_thread join: {}", err);
+    }
+    for hai in handlers {
+        if let Err(err) = hai.await {
+            warn!("benchmark handler_thread join: {}", err);
+        }
+    }
+}
+
+// once every 0.1s log a message for something that may be happening 10_000 times per second
+const BLAB_MICROS: i64 = 100_000;
+
+pub async fn direct_sender(
+    node_config: NodeConfig,
+    network_client: NetworkClient<BenchmarkMessage>,
+    time_service: TimeService,
+    network_id: NetworkId,
+    peer_id: PeerId,
+    shared: Arc<RwLock<BenchmarkSharedState>>,
+) {
+    let config = node_config.benchmark.unwrap();
+    let interval = Duration::from_nanos(1_000_000_000 / config.direct_send_per_second);
+    let ticker = time_service.interval(interval);
+    futures::pin_mut!(ticker);
+    // random payload filler
+    let data_size = config.direct_send_data_size;
+    let mut blob = Vec::<u8>::with_capacity(data_size);
+    let mut rng = OsRng;
+    for _ in 0..data_size {
+        blob.push(rng.gen());
+    }
+
+    let mut counter: u64 = rng.gen();
+
+    let mut next_blab: i64 = 0;
+
+    loop {
+        ticker.next().await;
+
+        counter += 1;
+        {
+            // tweak the random payload a little on every send
```

Comment:
What's the reason for this? Just to make things more "fair"? 😄

Reply:
I wanted to make sure compression and encryption had some work to do. An early version left the buffer zero filled and it compressed away entirely.